src/contracts/wrappers/ERC1155FloorWrapper.sol

```diff
@@ -0,0 +1,149 @@
+// SPDX-License-Identifier: Apache-2.0
+pragma solidity ^0.8.4;
+
+import {IERC165} from "@0xsequence/erc-1155/contracts/interfaces/IERC165.sol";
+import {IERC1155} from "@0xsequence/erc-1155/contracts/interfaces/IERC1155.sol";
+import {IERC1155TokenReceiver} from "@0xsequence/erc-1155/contracts/interfaces/IERC1155TokenReceiver.sol";
+import {ERC1155} from "@0xsequence/erc-1155/contracts/tokens/ERC1155/ERC1155.sol";
+import {ERC1155MintBurn} from "@0xsequence/erc-1155/contracts/tokens/ERC1155/ERC1155MintBurn.sol";
+import {IERC1155FloorWrapper} from "../interfaces/IERC1155FloorWrapper.sol";
+import {IERC1155Metadata} from "../interfaces/IERC1155Metadata.sol";
+import {IDelegatedERC1155Metadata} from "../interfaces/IDelegatedERC1155Metadata.sol";
+import {WrapperErrors} from "../utils/WrapperErrors.sol";
+
+/**
+ * @notice Allows users to wrap any amount of a ERC-1155 token with a 1:1 ratio.
+ * Therefore each ERC-1155 within a collection can be treated as if fungible.
+ */
+contract ERC1155FloorWrapper is
+    IERC1155FloorWrapper,
+    ERC1155MintBurn,
+    IERC1155Metadata,
+    IERC1155TokenReceiver,
+    WrapperErrors
+{
+    bool private initialized;
+    address internal immutable factory;
+    IERC1155 public token;
+    // This contract only supports a single token id
+    uint256 public constant TOKEN_ID = 0;
+
+    bool private isDepositing;
+
+    constructor() {
+        factory = msg.sender;
+    }
+
+    function initialize(address tokenAddr) external {
+        if (initialized || msg.sender != factory) {
+            revert InvalidInitialization();
+        }
+        token = IERC1155(tokenAddr);
+        initialized = true;
+    }
+
+    modifier onlyDepositing() {
+        if (!isDepositing) {
+            revert InvalidERC1155Received();
+        }
+        delete isDepositing;
```

Comment:
Why is this variable deleted twice?

Reply:
In case the ERC1155 doesn't correctly implement the callback. Having it cleared in the modifier handles duplicate receiver calls. Having it cleared in `deposit` prevents the flag being stuck on.

Reply:
Then we should remove it from the modifier (also having side effects on something that's called `onlyX` is not good). Are there many instances out there of ERC1155s not implementing the callback? I think the callback is way too fundamental to skip it.

Reply:
If we move from a `pull -> push` architecture this issue goes away anyway.

Reply:
Will keep this discussion here: https://github.com/0xsequence/niftyswap/pull/79#discussion_r1153674773